app/src/main/java/com/infomaniak/swisstransfer/ui/screen/newtransfer/TransferSendManager.kt

```diff
@@ -53,58 +53,100 @@ class TransferSendManager @Inject constructor(
-                    onFailure = ::setFailedIntegrityResult,
-                )
-            }.onFailure { exception ->
-                SentryLog.e(TAG, "Failed to start the upload", exception)
+
+        withIntegrityToken(
+            onSuccess = { attestationToken ->
+                sendTransfer(newUploadSession, attestationToken)
+            },
+            onRefused = {
+                _integrityCheckResult.value = AppIntegrityResult.Fail
+            },
+            onFailure = { exception ->
+                if (exception !is CancellationException) {
+                    SentryLog.e(TAG, "Integrity token received an exception", exception)
+                } else {
+                    SentryLog.i(TAG, "Integrity token received an exception", exception)
+                }
                 _sendActionResult.update { SendActionResult.Failure }
             }
+        )
+    }
+
+    private suspend fun sendTransfer(newUploadSession: NewUploadSession, attestationToken: String) {
+        _integrityCheckResult.value = AppIntegrityResult.Success
+        _sendActionResult.update { SendActionResult.Pending }
+
+        runCatching {
+            val uuid = uploadManager.createAndGetUpload(newUploadSession).uuid
+            uploadManager.initUploadSession(
+                attestationHeaderName = AppIntegrityManager.ATTESTATION_TOKEN_HEADER,
+                attestationToken = attestationToken,
+            )!! // TODO: Handle ContainerErrorsException here
+            uploadWorkerScheduler.scheduleWork(uuid)
+            _sendActionResult.update {
+                val totalSize = importationFilesManager.importedFiles.value.sumOf { it.fileSize }
+                SendActionResult.Success(totalSize)
+            }
+        }.onFailure { exception ->
+            SentryLog.e(TAG, "Failed to start the upload", exception)
+            _sendActionResult.update { SendActionResult.Failure }
         }
     }
 
-    private fun CoroutineScope.requestAppIntegrityToken(
-        appIntegrityManager: AppIntegrityManager,
-        newUploadSession: NewUploadSession,
+    //region App Integrity
+    private suspend inline fun withIntegrityToken(
+        onSuccess: (attestationToken: String) -> Unit,
+        onRefused: () -> Unit = {},
+        onFailure: (exception: Throwable) -> Unit = {},
     ) {
+        runCatching {
+            var attestationToken: String? = null
+
+            coroutineScope {
+                appIntegrityManager.getChallenge(
+                    onSuccess = { launch { attestationToken = requestAppIntegrityToken(appIntegrityManager) } },
+                    onFailure = {},
+                )
+            }
+
+            attestationToken?.let { onSuccess(it) } ?: onRefused.invoke()
+        }.onFailure {
+            onFailure.invoke(it)
+        }
+    }
+
+    private suspend fun requestAppIntegrityToken(appIntegrityManager: AppIntegrityManager): String? = coroutineScope {
+        var attestationToken: String? = null
+
```

Comment:
This won't work because the attestation won't be assign with the good value as it's inside the coroutineScope, so it won't wait until the `launch`completion. 
We need to place the variable outside of the scope in order to execute the launch before the assignation. 

```suggestion
    private suspend fun requestAppIntegrityToken(appIntegrityManager: AppIntegrityManager): String? {
        var attestationToken: String? = null

        coroutineScope {
```